scylla/src/transport/timestamp_generator.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    sync::atomic::AtomicI64,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use std::sync::atomic::Ordering;
+use std::sync::Mutex;
+use tokio::time::{Duration, Instant};
+use tracing::warn;
+
+/// Trait used to represent a timestamp generator
+pub trait TimestampGenerator: Send + Sync {
+    // This generates a new timestamp
+    fn next_timestamp(&self) -> i64;
+}
+
+/// Basic timestamp generator. Provides no guarantees, if system clock returns
+/// time before UNIX epoch it panics.
+#[derive(Default)]
+pub struct SimpleTimestampGenerator {}
+
+#[allow(dead_code)]
+impl SimpleTimestampGenerator {
+    pub fn new() -> Self {
+        SimpleTimestampGenerator {}
+    }
+}
+
+impl TimestampGenerator for SimpleTimestampGenerator {
+    fn next_timestamp(&self) -> i64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_micros() as i64
+    }
+}
+
+/// Warning configuration for MonotonicTimestampGenerator
+pub struct MonotonicTimestampGeneratorCfg {
+    warning_threshold: Duration,
+    warning_interval: Duration,
+}
+
+/// Monotonic timestamp generator. Guarantees monotonicity of timestamps.
+/// If system clock will not provide an increased timestamp, then the timestamp will
+/// be artificially increased. If the config is provided and the clock skew is bigger than
+/// warning_threshold (by default 1 second), then the user will be warned about
+/// the skew repeatedly, with warning_interval provided in the settings (by default 1 second).
+/// Remember that this generator only guarantees monotonicity within one client!
+/// If you create multiple clients with MonotonicTimestampGenerators the monotonicity
+/// guarantee becomes void.
+pub struct MonotonicTimestampGenerator {
+    last: AtomicI64,
+    last_warning: Mutex<Instant>,
+    config: Option<MonotonicTimestampGeneratorCfg>,
+}
+
+impl MonotonicTimestampGenerator {
+    /// Creates a new monotonic timestamp generator with provided settings
+    pub fn new_with_settings(config: Option<MonotonicTimestampGeneratorCfg>) -> Self {
+        MonotonicTimestampGenerator {
+            last: AtomicI64::new(0),
+            last_warning: Mutex::new(Instant::now()),
+            config,
+        }
+    }
```

Comment:
If we predict that the configuration options of the generator could be extended in the future, then making the config public together with exposing a set of configuration methods would be the better option.
OTOH, if it's unlikely for the generator to be extended, it would be better not to expose another struct (especially one with such a long name :wink: ).

Reply:
I don't think it's likely. And if it is likely, then we need another struct for config. The purpose of this one is a type level guarantee that either both interval and threshold are set, or none of them are.